### DISCUSSION TITLE
Is there a way to dynamically detect MSSQL Default constraints being set to a column?

### DISCUSSION THREAD

**Discussion by indiVar0508** [sqlalchemy/sqlalchemy#9346] [Usage Questions] [answered]
Hi,

Is there a way to detect default constraints for a column that is set by MSSQL `DF_TABLE_NAME_123456` ?

Thanks

**Comment by zzzeek** [comment]
hi - 

I can't answer this question because I have no idea what it refers towards.    

@gordthompson do you know what this is referring towards?

**Reply by gordthompson** [reply]
If we script a table in SSMS the column defaults show up as constraints:

```sql
CREATE TABLE [dbo].[my_table](
	[n] [int] NULL,
	[status] [varchar](50) NULL
) ON [PRIMARY]
GO

ALTER TABLE [dbo].[my_table] ADD  CONSTRAINT [DF_my_table_status]  DEFAULT ('unknown') FOR [status]
GO
```

**Comment by gordthompson** [comment, marked as answer]
You can use an [Inspector](https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector) to view the properties of the columns in a table. One of those properties is "default":

```python
insp = inspect(engine)
col_info = insp.get_columns("my_table")
pprint(col_info, sort_dicts=False)
"""
[{'name': 'n',
  'type': INTEGER(),
  'nullable': True,
  'default': None,
  'autoincrement': False,
  'comment': None},
 {'name': 'status',
  'type': VARCHAR(length=50, collation='SQL_Latin1_General_CP1_CI_AS'),
  'nullable': True,
  'default': "('unknown')",
  'autoincrement': False,
  'comment': None}]
"""
```

**Reply by indiVar0508** [reply]
Thanks @gordthompson , 
this shows default value, actually in my case i wanted to detect the name `DF_TABLE_NAME_123456` itself as,
currently i am doing testing for migration of DB with my APP, so when i am downgrading my DB with MSSQL DB i get error saying `The object 'DF__table_nam__col_n__53BC6B19' is dependent on column 'col_name'` when altering column and i can't hardcode the constraint name as every upgrade seems to be generating new name. So wanted to see if there is a way to get name so i can drop this constraint when altering the column and create a new contraint after altering. asked same in alembic also 
 https://github.com/sqlalchemy/alembic/discussions/1183

For now, for a workaround i am creating backup of data column, dropping the column and creating new column and loading backup data to make it work.

**Reply by gordthompson** [reply]
You can retrieve the names of the default constraints with a query like this:

```sql
select sch.name as schema_name, tbl.name as table_name, 
    col.name as column_name, def.name as constraint_name
from sys.default_constraints def 
    inner join sys.tables tbl 
        on def.parent_object_id = tbl.object_id 
    inner join sys.schemas sch
        on tbl.schema_id = sch.schema_id
    inner join sys.columns col
        on def.parent_object_id = col.object_id
            and def.parent_column_id = col.column_id
```

**Reply by indiVar0508** [reply]
Thanks @gordthompson , 
this works for me ! :D
